server.py

```diff
@@ -1,36 +1,376 @@
+            raise KeyError(f"Coach ID '{coach.id}' not found in DB")
+        self._db[coach.id] = coach
+        logger.debug("Updated coach with ID %d", coach.id)
+        if self._auto_persist:
+            self.persist(self._file)
+        return coach
+
+    def persist(self, file: os.PathLike) -> None:
+        """Persist the database to a given file."""
+        logger.debug("Saving coaches DB to file: %s", file)
+        data = dict(next_id=self._next_id, data=[c.to_json() for c in self])
+        with open(file, "w") as f:
+            json.dump(data, f)
+
+    def _load_from_file(self, file: os.PathLike):
+        """Load data in from a given file, or create the file if it doesn't exist."""
+        try:
+            with open(file, "r") as f:
+                data = json.load(f)
+            logger.info("Loading in coaches DB from file: %s", file)
+            self._db = {c["id"]: Coach.from_json(c) for c in data["data"]}
+            self._next_id = data["next_id"]
+        except FileNotFoundError:
+            logger.info("Coaches DB file not found, creating now: %s", file)
+            self.persist(file)
+
+
+# The in-memory database of coaches.
+coach_db: CoachDB
+
+# The file used to persist the coach DB.
+COACH_DB_FILE = "./coach_db.json"
+
+
+# ------------------------------------------------------------------------------
+# Helpers
+# ------------------------------------------------------------------------------
+
+
+def _parse_languages(languages: str) -> Dict[str, int]:
+    ret = dict()
+    if not languages:
+        return ret
+    for lang in languages.split(","):
+        lang_name, proficiency = lang.split(":")
+        ret[lang_name] = int(proficiency)
+    return ret
+
+
+# ------------------------------------------------------------------------------
+# Route handlers
+# ------------------------------------------------------------------------------
+
+
+def get_coaches() -> List[Coach]:
+    return list(coach_db)
+
+
+def get_coach(coach_id: int) -> Coach:
+    return coach_db.get(coach_id)
+
+
+def create_coach(data: CoachJSON) -> Coach:
+    coach = Coach(**data)
+    db_coach = coach_db.add(coach)
+    return db_coach
+
+
+def edit_coach(coach_id: int, data: CoachJSON) -> Coach:
+    kwargs = get_coach(coach_id).to_json()
+    kwargs.update(data)
+    updated_coach = Coach(**kwargs)
+    db_coach = coach_db.update_entry(updated_coach)
+    return db_coach
+
+
+def delete_coach(coach_id: int) -> None:
+    coach_db.remove(coach_id)
+
+
+def get_coach_matches(data: CoachJSON) -> List[Tuple[int, Coach]]:
+    """
+    Get coaches based on how well they match the given data.
+
+    :param data:
+        Data to match on.
+    :return:
+        A list of tuples containing the match score and the matched coach.
+    """
+
+    def get_lang_match_score(
+        c: Coach, languages: Dict[str, int]
+    ) -> Optional[Tuple[str, int]]:
+        match_langs = [L for L in languages if L in c.languages]
```

Comment:
Doesn't really matter, but not a big fan of upper-case single-character variables :)

Reply:
This is a standard special-case used for 'L' since the lowercase variant is awful (could be capital 'i' or '1').